File: crates/juicefront/ui/src/lib/metadata.ts
import { execSync } from "child_process";

export interface RepoMetadata {
  owner: string;
  name: string;
  defaultBranch: string;
  url: string;
  latestCommitRef: string;
  latestCommitHash: string;
  latestCommitShortHash: string;
}

function exec(cmd: string): string {
  try {
    return execSync(cmd, { cwd: process.cwd() }).toString().trim();
  } catch {
    return "unknown";
  }
}

const gitRef = exec("git rev-parse --abbrev-ref HEAD 2>/dev/null");
const gitHash = exec("git rev-parse HEAD 2>/dev/null");
const gitShortHash = gitHash !== "unknown" ? gitHash.slice(0, 7) : "unknown";

export const REPO_METADATA: RepoMetadata = {
  owner: "create-juicey-app",
  name: "juicebox2",
  defaultBranch: "main",
  url: "https://github.com/create-juicey-app/juicebox2",
  latestCommitRef: gitRef,
  latestCommitHash: gitHash,
  latestCommitShortHash: gitShortHash,
};

export const getRepoUrl = (): string => REPO_METADATA.url;

export const getRepoPathUrl = (
  path: string,
  ref: string = REPO_METADATA.defaultBranch,
): string =>
  `${REPO_METADATA.url}/blob/${encodeURIComponent(ref)}/${path.replace(/^\/+/, "")}`;

export const getFormattedCommitLabel = (includeBranch = true): string =>
  includeBranch
    ? REPO_METADATA.latestCommitRef
    : REPO_METADATA.latestCommitShortHash;

export const getProjectStructuredData = () => ({
  "@context": "https://schema.org",
  "@type": "SoftwareSourceCode",
  name: "Juicebox2",
  version: REPO_METADATA.latestCommitShortHash,
  programmingLanguage: "TypeScript",
  codeRepository: REPO_METADATA.url,
  license: "https://opensource.org/licenses/MIT",
  source: {
    "@type": "URL",
    "@id": REPO_METADATA.url,
  },
});

File: crates/juicefront/src/ssr/metadata.rs
pub const REPO_URL: &str = "https://github.com/create-juicey-app/juicebox2";

pub fn commit_ref() -> &'static str {
    option_env!("JUICEFRONT_COMMIT_REF").unwrap_or("unknown")
}

pub fn commit_hash() -> &'static str {
    option_env!("JUICEFRONT_COMMIT_HASH").unwrap_or("unknown")
}

pub fn commit_short_hash() -> String {
    let hash = commit_hash();
    if hash == "unknown" {
        return "unknown".to_owned();
    }
    hash.chars().take(7).collect()
}

pub fn commit_label() -> String {
    commit_ref().to_owned()
}

pub fn structured_data_json() -> String {
    serde_json::json!({
        "@context": "https://schema.org",
        "@type": "SoftwareSourceCode",
        "name": "Juicebox2",
        "version": commit_short_hash(),
        "programmingLanguage": "Rust",
        "codeRepository": REPO_URL,
        "license": "https://opensource.org/licenses/MIT",
        "source": {
            "@type": "URL",
            "@id": REPO_URL,
        },
    })
    .to_string()
}
